File: TeamCode/Proto/src/main/java/org/murraybridgebunyips/proto/Proto.java
package org.murraybridgebunyips.proto;

import static org.murraybridgebunyips.bunyipslib.external.units.Units.DegreesPerSecond;
import static org.murraybridgebunyips.bunyipslib.external.units.Units.Inches;
import static org.murraybridgebunyips.bunyipslib.external.units.Units.InchesPerSecond;
import static org.murraybridgebunyips.bunyipslib.external.units.Units.Millimeters;
import static org.murraybridgebunyips.bunyipslib.external.units.Units.Second;

import com.acmerobotics.roadrunner.control.PIDCoefficients;
import com.qualcomm.hardware.rev.RevHubOrientationOnRobot;
import com.qualcomm.robotcore.hardware.DcMotorSimple;
import com.qualcomm.robotcore.hardware.IMU;

import org.murraybridgebunyips.bunyipslib.Dbg;
import org.murraybridgebunyips.bunyipslib.Motor;
import org.murraybridgebunyips.bunyipslib.RobotConfig;
import org.murraybridgebunyips.bunyipslib.roadrunner.drive.DriveConstants;
import org.murraybridgebunyips.bunyipslib.roadrunner.drive.MecanumCoefficients;
import org.murraybridgebunyips.bunyipslib.roadrunner.drive.localizers.TwoWheelLocalizer;
import org.murraybridgebunyips.bunyipslib.roadrunner.util.Deadwheel;

/**
 * FTC 15215 INTO THE DEEP 2024-2025 robot configuration
 *
 * @author Lucas Bubner, 2024
 */
public class Proto extends RobotConfig {
    /**
     * Expansion 0: Front Left "fl"
     */
    public Motor frontLeft;
    /**
     * Expansion 1: Front Right "fr"
     */
    public Motor frontRight;
    /**
     * Expansion 2: Back Right "br"
     */
    public Motor backRight;
    /**
     * Expansion 3: Back Left "bl"
     */
    public Motor backLeft;
    /**
     * Control 3: Parallel Encoder "pe"
     */
    public Deadwheel parallelDeadwheel;
    /**
     * Control 2: Perpendicular Encoder "ppe"
     */
    public Deadwheel perpendicularDeadwheel;
    /**
     * Internally mounted on I2C C0 "imu"
     */
    public IMU imu;

    /**
     * RoadRunner drive constants
     */
    public DriveConstants driveConstants;
    /**
     * Dual deadwheel intrinsics
     */
    public TwoWheelLocalizer.Coefficients localizerCoefficients;
    /**
     * Mecanum coefficients
     */
    public MecanumCoefficients mecanumCoefficients;

    @Override
    protected void onRuntime() {
        // Base is from GLaDOS
        imu = getHardware("imu", IMU.class, (d) -> {
            boolean init = d.initialize(new IMU.Parameters(
                    new RevHubOrientationOnRobot(
                            RevHubOrientationOnRobot.LogoFacingDirection.UP,
                            RevHubOrientationOnRobot.UsbFacingDirection.LEFT
                    )
            ));
            if (!init) Dbg.error("imu failed init");
        });
        frontLeft = getHardware("fl", Motor.class, (d) -> d.setDirection(DcMotorSimple.Direction.FORWARD));
        frontRight = getHardware("fr", Motor.class, (d) -> d.setDirection(DcMotorSimple.Direction.REVERSE));
        backRight = getHardware("br", Motor.class, (d) -> d.setDirection(DcMotorSimple.Direction.REVERSE));
        backLeft = getHardware("bl", Motor.class, (d) -> d.setDirection(DcMotorSimple.Direction.REVERSE));

        // REV Through Bore Encoders
        parallelDeadwheel = getHardware("pe", Deadwheel.class,
                (d) -> d.setDirection(Deadwheel.Direction.REVERSE));
        perpendicularDeadwheel = getHardware("ppe", Deadwheel.class,
                (d) -> d.setDirection(Deadwheel.Direction.FORWARD));
        localizerCoefficients = new TwoWheelLocalizer.Coefficients.Builder()
                .setTicksPerRev(8192)
                .setGearRatio(1)
                .setWheelRadius(Millimeters.of(60).divide(2))
                .setXMultiplier(70.0 / 69.1)
                .setYMultiplier(70.0 / 69.1)
                .setParallelX(Inches.zero())
                .setParallelY(Inches.of(-1.8))
                .setPerpendicularX(Inches.one())
                .setPerpendicularY(Inches.of(1.7))
                .setOverflowCompensation(true)
                .build();

        // RoadRunner configuration
        driveConstants = new DriveConstants.Builder()
                .setTicksPerRev(28)
                .setMaxRPM(6000)
                .setRunUsingEncoder(false)
                .setWheelRadius(Millimeters.of(75).divide(2))
                .setGearRatio(1.0 / 13.1)
                .setTrackWidth(Inches.of(20.5))
                .setMaxVel(InchesPerSecond.of(48))
                .setMaxAccel(InchesPerSecond.per(Second).of(96))
                .setMaxAngVel(DegreesPerSecond.of(175))
                .setMaxAngAccel(DegreesPerSecond.per(Second).of(175))
                .setKV(0.01395)
                .setKStatic(0.06311)
                .setKA(0.0015)
                .build();
        mecanumCoefficients = new MecanumCoefficients.Builder()
                .setLateralMultiplier(60.0 / 59.846666)
                .setTranslationalPID(new PIDCoefficients(8, 0, 0))
                .setHeadingPID(new PIDCoefficients(10, 0, 0))
                .build();
    }
}

File: TeamCode/Proto/src/main/java/au/edu/sa/mbhs/studentrobotics/ftc15215/proto/Proto.kt
package au.edu.sa.mbhs.studentrobotics.ftc15215.proto

import au.edu.sa.mbhs.studentrobotics.bunyipslib.BunyipsOpMode
import au.edu.sa.mbhs.studentrobotics.bunyipslib.RobotConfig
import au.edu.sa.mbhs.studentrobotics.bunyipslib.external.control.TrapezoidProfile
import au.edu.sa.mbhs.studentrobotics.bunyipslib.external.control.ff.ElevatorFeedforward
import au.edu.sa.mbhs.studentrobotics.bunyipslib.external.control.pid.PController
import au.edu.sa.mbhs.studentrobotics.bunyipslib.external.units.Unit.Companion.of
import au.edu.sa.mbhs.studentrobotics.bunyipslib.external.units.Units.Inches
import au.edu.sa.mbhs.studentrobotics.bunyipslib.external.units.Units.InchesPerSecond
import au.edu.sa.mbhs.studentrobotics.bunyipslib.external.units.Units.InchesPerSecondPerSecond
import au.edu.sa.mbhs.studentrobotics.bunyipslib.external.units.Units.Milliseconds
import au.edu.sa.mbhs.studentrobotics.bunyipslib.external.units.Units.Seconds
import au.edu.sa.mbhs.studentrobotics.bunyipslib.hardware.IMUEx
import au.edu.sa.mbhs.studentrobotics.bunyipslib.hardware.Motor
import au.edu.sa.mbhs.studentrobotics.bunyipslib.hardware.ServoEx
import au.edu.sa.mbhs.studentrobotics.bunyipslib.hardware.SimpleRotator
import au.edu.sa.mbhs.studentrobotics.bunyipslib.hooks.BunyipsLib
import au.edu.sa.mbhs.studentrobotics.bunyipslib.localization.TwoWheelLocalizer
import au.edu.sa.mbhs.studentrobotics.bunyipslib.roadrunner.parameters.DriveModel
import au.edu.sa.mbhs.studentrobotics.bunyipslib.roadrunner.parameters.MecanumGains
import au.edu.sa.mbhs.studentrobotics.bunyipslib.roadrunner.parameters.MotionProfile
import au.edu.sa.mbhs.studentrobotics.bunyipslib.subsystems.HoldableActuator
import au.edu.sa.mbhs.studentrobotics.bunyipslib.subsystems.Switch
import au.edu.sa.mbhs.studentrobotics.bunyipslib.subsystems.drive.MecanumDrive
import au.edu.sa.mbhs.studentrobotics.bunyipslib.tasks.bases.Task.Companion.task
import au.edu.sa.mbhs.studentrobotics.bunyipslib.vision.Vision
import com.acmerobotics.roadrunner.ftc.RawEncoder
import com.qualcomm.hardware.rev.RevHubOrientationOnRobot
import com.qualcomm.robotcore.eventloop.opmode.Autonomous
import com.qualcomm.robotcore.hardware.DcMotorEx
import com.qualcomm.robotcore.hardware.DcMotorSimple
import com.qualcomm.robotcore.hardware.IMU
import com.qualcomm.robotcore.hardware.Servo
import com.qualcomm.robotcore.hardware.TouchSensor
import org.firstinspires.ftc.robotcore.external.hardware.camera.WebcamName

/**
 * FTC 15215 INTO THE DEEP 2024-2025 robot configuration
 *
 * @author Lucas Bubner, 2024
 */
@RobotConfig.AutoInit
object Proto : RobotConfig() {
    /**
     * Non-subsystem hardware instances.
     */
    val hw = Hardware()

    /**
     * RoadRunner Mecanum Drive with Two-Wheel Localization.
     */
    lateinit var drive: MecanumDrive

    /**
     * `claws` rotation mechanism.
     */
    lateinit var clawRotator: Switch

    /**
     * Vertical lift for the claw mechanism (`claws` and `clawRotator`).
     */
    lateinit var clawLift: HoldableActuator

    /**
     * Forward camera.
     */
    lateinit var camera: Vision

    val eject = {
        task {
            timeout(500 of Milliseconds)
            init {
                // Negative ejects
                hw.clawIntake?.power = -1.0
            }
            onFinish {
                hw.clawIntake?.power = 0.0
            }
        }
    }

    override fun onRuntime() {
        // Base is from GLaDOS
        hw.imu = getHardware("imu", IMUEx::class.java) {
            it.lazyInitialize(IMU.Parameters(RevHubOrientationOnRobot(
                RevHubOrientationOnRobot.LogoFacingDirection.UP,
                RevHubOrientationOnRobot.UsbFacingDirection.LEFT
            )))
        }
        hw.fl = getHardware("fl", Motor::class.java) {
            it.direction = DcMotorSimple.Direction.FORWARD
            it.setPowerDeltaThreshold(0.02)
        }
        hw.bl = getHardware("bl", Motor::class.java) {
            it.direction = DcMotorSimple.Direction.FORWARD
            it.setPowerDeltaThreshold(0.02)
        }
        hw.br = getHardware("br", Motor::class.java) {
            it.direction = DcMotorSimple.Direction.REVERSE
            it.setPowerDeltaThreshold(0.02)
        }
        hw.fr = getHardware("fr", Motor::class.java) {
            it.direction = DcMotorSimple.Direction.FORWARD
            it.setPowerDeltaThreshold(0.02)
        }

        // REV Through Bore Encoders attached on ports 0 and 3 with the motors
        hw.pe = getHardware("fr", RawEncoder::class.java) {
            it.direction = DcMotorSimple.Direction.FORWARD
        }
        hw.ppe = getHardware("br", RawEncoder::class.java) {
            it.direction = DcMotorSimple.Direction.REVERSE
        }

        // End effectors
        hw.clawIntake = getHardware("cs", SimpleRotator::class.java) {
            it.setPowerDeltaThreshold(0.02)
        }
        hw.clawRotator = getHardware("cr", ServoEx::class.java) {
            it.setConstraints(TrapezoidProfile.Constraints(Constants.cr_v, Constants.cr_a))
            it.direction = Servo.Direction.REVERSE
            it.setPositionDeltaThreshold(0.02)
            it.scaleRange(0.25, 1.0)
        }

        hw.clawLift = getHardware("cl", Motor::class.java) {
            it.direction = DcMotorSimple.Direction.REVERSE
            val p = PController(Constants.cl_kP)
            val ff = ElevatorFeedforward(0.0, Constants.cl_kG, 0.0, 0.0, { 0.0 }, { 0.0 })
            val c = p.compose(ff) { a, b -> a + b }
            it.runToPositionController = c
            it.setPowerDeltaThreshold(0.02)
            BunyipsOpMode.ifRunning { o ->
                o.onActiveLoop {
                    c.setCoefficients(
                        Constants.cl_kP,
                        0.0,
                        0.0,
                        0.0,
                        0.0,
                        Constants.cl_kG,
                        0.0,
                        0.0
                    )
                }
            }
        }
        hw.bottom = getHardware("bottom", TouchSensor::class.java)

        hw.camera = getHardware("webcam", WebcamName::class.java)

        // RoadRunner drivebase configuration
        val dm = DriveModel.Builder()
            .setDistPerTick(100.0 of Inches, 109586.0) // 0.000912525
            .setLateralInPerTick(0.0007272580236408389)
            .setTrackWidthTicks(15504.984641208961)
            .build()
        val mp = MotionProfile.Builder()
            .setMaxWheelVel(50.0 of InchesPerSecond)
            .setMinProfileAccel(-80.0 of InchesPerSecondPerSecond)
            .setMaxProfileAccel(80.0 of InchesPerSecondPerSecond)
            .setKs(1.182874620902731)
            .setKv(0.0002)
            .setKa(0.000032)
            .build()
        val mg = MecanumGains.Builder()
            .setAxialGain(2.5)
            .setLateralGain(2.5)
            .setHeadingGain(4.0)
            .build()
        val twl = TwoWheelLocalizer.Params.Builder()
            .setParYTicks(-2331.640217956543)
            .setPerpXTicks(-6317.573653757131)
            .build()

//        val at = AprilTag {
//            AprilTag.setCameraPose(it)
//                .forward(Inches.of(8.0))
//                .right(Inches.one())
//                .apply()
//        }
        camera = Vision(hw.camera)
        drive = MecanumDrive(dm, mp, mg, hw.fl, hw.bl, hw.br, hw.fr, hw.imu as IMU, hardwareMap.voltageSensor)
            .withLocalizer(TwoWheelLocalizer(dm, twl, hw.pe, hw.ppe, hw.imu?.get()))
            .withName("Drive")
        clawRotator = Switch(hw.clawRotator)
            .withName("Claw Rotator")
        clawLift = HoldableActuator(hw.clawLift)
            .withBottomSwitch(hw.bottom)
            .withUserSetpointControl { dt -> dt * Constants.cl_TPS }
            .withMaxSteadyStateTime(10 of Seconds)
            .withUpperLimit(Constants.cl_MAX)
            .withName("Claw Lift")

        if (BunyipsLib.opMode.javaClass.isAnnotationPresent(Autonomous::class.java)) {
            BunyipsOpMode.instance.setInitTask(clawLift.tasks.home().mutate().addPeriodic { clawLift.update() })
            clawLift.withTolerance(25)
        }

//        BunyipsOpMode.ifRunning {
//            it.onActiveLoop {
//                Motor.debug(hw.clawLift!!, "Vertical Lift", it.t)
//            }
//        }
    }

    /**
     * Raw hardware instances.
     */
    class Hardware {
        /**
         * Expansion 0: Front Left "fl"
         */
        var fl: DcMotorEx? = null

        /**
         * Expansion 1: Front Right "fr"
         */
        var fr: DcMotorEx? = null

        /**
         * Expansion 2: Back Right "br"
         */
        var br: DcMotorEx? = null

        /**
         * Expansion 3: Back Left "bl"
         */
        var bl: DcMotorEx? = null

        /**
         * Expansion 3: Parallel Encoder "fr" (on motor)
         */
        var pe: RawEncoder? = null

        /**
         * Expansion 0: Perpendicular Encoder "br" (on motor)
         */
        var ppe: RawEncoder? = null

        /**
         * Internally mounted on I2C C0 "imu"
         */
        var imu: IMUEx? = null

        /**
         * Control S1: Claw Spinny "cs"
         */
        var clawIntake: SimpleRotator? = null

        /**
         * Control S2: Claw Rotator "cr"
         */
        var clawRotator: Servo? = null

        /**
         * Control 1: Claw Lift "cl"
         */
        var clawLift: Motor? = null

        /**
         * Control Digital 1: "bottom" limit for claw lift
         */
        var bottom: TouchSensor? = null

        /**
         * Control USB 3.0: Webcam
         */
        var camera: WebcamName? = null
    }
}